src/litserve/specs/openai.py

```diff
@@ -130,68 +134,93 @@ def encode_response(self, output: Union[Dict[str, str], List[Dict[str, str]]]) -
+        return choice_pipes
+
+    async def options_chat_completions(self, request: Request):
+        return Response(status_code=200)
 
-    async def chat_completion(
-        self, request: ChatCompletionRequest, background_tasks: BackgroundTasks
-    ) -> ChatCompletionResponse:
+    async def chat_completion(self, request: ChatCompletionRequest, background_tasks: BackgroundTasks):
         logger.debug("Received chat completion request %s", request)
-        if request.stream:
-            raise HTTPException(400, "Stream is not supported")
 
         uids = [uuid.uuid4() for _ in range(request.n)]
         pipes = []
         for uid in uids:
             read, write = self._server.new_pipe()
-
             request_el = request.model_copy()
             request_el.n = 1
             self._server.request_buffer[uid] = (request_el, write)
             self._server.request_queue.put(uid)
-
             background_tasks.add_task(self._server.cleanup_request, self._server.request_buffer, uid)
             pipes.append((read, write))
 
-        responses = await self.get_from_pipe(uids, pipes)
-
+        responses = await self.get_from_pipes(uids, pipes)
         for read, write in pipes:
             self._server.dispose_pipe(read, write)
 
-        usage = UsageInfo()
+        if request.stream:
+            return StreamingResponse(self.streaming_completion(request, responses), media_type="application/x-ndjson")
 
-        choices = []
-        for i, data in enumerate(responses):
-            response, status = data
-            logger.debug("Received chat completion response %s with status %s", response, status)
-            if status == LitAPIStatus.ERROR:
-                load_and_raise(response)
+        return await self.non_streaming_completion(request, responses)
 
-            if status != LitAPIStatus.OK:
-                break
+    async def streaming_completion(self, request: ChatCompletionRequest, pipe_responses: List):
+        model = request.model
+        usage = None
+        async for streaming_response in azip(*pipe_responses):
+            choices = []
+            for i, chat_msg in enumerate(streaming_response):
+                chat_msg = json.loads(chat_msg)
+                logger.debug(chat_msg)
+                chat_msg = ChoiceDelta(**chat_msg)
+                choice = ChatCompletionStreamingChoice(
+                    index=i, delta=chat_msg, system_fingerprint="", usage=usage, finish_reason=None
+                )
+
+                choices.append(choice)
+
+            chunk = ChatCompletionChunk(model=model, choices=choices, usage=usage).json()
+            logger.debug(chunk)
+            yield f"data: {chunk}\n\n"
 
-            response = response.model_copy(update={"index": i})
-            choices.append(response)
+        last_chunk = ChatCompletionChunk(
+            model=model,
+            choices=[ChatCompletionStreamingChoice(index=0, delta=ChoiceDelta(), finish_reason="stop")],
```

Comment:
right but even in this case we can't have just `index=0`, or anything consuming this will break
we need to create an empty `ChoiceDelta` for every choice (i.e. index=0..len(choices))

Reply:
oh I see your point. so when `request.n>1` we need to have that many empty choices, right? How about - 

```py
choices = [ChatCompletionStreamingChoice(index=i, delta=ChoiceDelta(), finish_reason="stop") for i in range(request.n)]
        last_chunk = ChatCompletionChunk(
            model=model,
            choices=choices,
            usage=usage,
        ).json()
        yield f"data: {last_chunk}\n\n"
        yield "data: [DONE]\n\n"
```